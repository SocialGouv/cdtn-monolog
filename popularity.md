# Popularity

Stats defined in [popularity.js](src/analysis/popularity.js)
The report Popularity compute the evolution of unique page views compared with a reference.

The page views is normalized as % of the total content 
- normalize = #page_views/#Total_views_for_the_period
```js
{
  'focus_count': # of views of the page on the focus period
  'ref_count': # of views of the page on the reference period
  'diff': focus_norm_count - row.ref_norm_count (variation of % points between focus and ref)
  'abs_diff': abs(diff) (variation up or down)
}
```



