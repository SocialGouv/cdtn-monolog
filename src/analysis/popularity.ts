import { DataFrame, IDataFrame } from "data-forge";
import { isSome, none, Option } from "fp-ts/lib/Option";

import { Cache } from "../cdtn/cdtn.types";
import { getVisits, toUniqueSearches, toUniqueViews } from "../reader/dataset";
import { actionTypes, urlToPath } from "../reader/readerUtil";
import { PopularityReport } from "./reports.types";

const reportType = (pt: PopularityTypeString): string =>
  `${pt.toLowerCase()}-popularity`;

const computeReports = (
  focusCounts: IDataFrame,
  refCounts: IDataFrame,
  previousMonthCount: IDataFrame,

  focusStart: number,
  refStart: number,
  prevStart: number,

  reportId: string,
  reportType: string
) => {
  // FIXME use outer join to handle missing values (e.g. additions)

  const joinedM0M1 = refCounts.join(
    focusCounts,
    (left) => left.field,
    (right) => right.field,
    (left, right) => {
      return {
        field: left.field,
        m0_count: right.count,
        m0_norm_count: right.normalized_count,
        m1_count: left.count,
        m1_norm_count: left.normalized_count,
      };
    }
  );

  const joined = joinedM0M1.join(
    previousMonthCount,
    (left) => left.field,
    (right) => right.field,
    (left, right) => {
      return {
        ...left,
        m2_count: right.count,
      };
    }
  );

  const nContent = 40;
  const minOccurence = 40;

  const diff = joined
    .generateSeries({
      diff: (row) => row.m0_norm_count - row.m1_norm_count,
    })
    .generateSeries({
      abs_diff: (row) => Math.abs(row.diff),
    })
    .generateSeries({
      rel_diff: (row) => (row.m0_count - row.m1_count) / row.m1_count,
    })
    .where((r) => r.m1_count + r.m0_count + r.m2_count > minOccurence);

  const topDiff = diff.orderByDescending((r: any) => r.abs_diff).take(nContent);
  const topPop = diff.orderByDescending((r) => r.m0_count).take(nContent);

  const top = topDiff.concat(topPop).distinct((row) => row.field);

  return top.toArray().map((doc) => ({
    doc,
    m0_start: focusStart * 1000,
    m1_start: refStart * 1000,
    m2_start: prevStart * 1000,
    reportId,
    reportType,
  }));
};

export enum PopularityType {
  CONTENT,
  CONVENTION,
  QUERY,
}

type PopularityTypeString = keyof typeof PopularityType;

const countURLs = (dataframe: IDataFrame) => {
  const counts = dataframe
    .deflate((a) => a.url)
    .groupBy((value) => value)
    .select((group) => {
      return {
        count: group.count(),
        field: group.first(),
      };
    })
    .inflate()
    .orderByDescending((r) => r.count);

  const sum = counts.deflate((r) => r.count).sum();
  const normalizedCounts = counts.withSeries({
    normalized_count: (df) =>
      df.deflate((row) => row.count).select((count) => count / sum),
  });
  return normalizedCounts.setIndex("url");
};

export const countQueries = (
  logs: IDataFrame,
  cache: Option<Cache>
): DataFrame => {
  if (isSome(cache)) {
    const queries = logs
      .where((a) => a.type == actionTypes.search)
      .getSeries("query")
      .toArray()
      .filter((a) => a && a != "undefined")
      .map((q) => q.toLowerCase());

    type ClusterCount = Array<{ query: string; count: number }>;
    const clusterCounts = new Map<number, ClusterCount>();

    console.log(queries.length);

    queries.forEach((currQ) => {
      const idx = cache.value.queryMap.get(currQ);
      const entry = idx && cache.value.clusters.get(idx);
      if (!idx || !entry) {
        //TODO
        //   console.log("Issue " + query);
        //   console.log(idx);
        return;
      }

      if (!clusterCounts.has(idx)) {
        // add cluster
        const cluster = entry.queries.map((query) => ({ count: 0, query }));
        clusterCounts.set(idx, cluster);
      }

      // increment count
      const cc = clusterCounts.get(idx)?.find(({ query }) => currQ == query);
      if (!cc) {
        //TODO
        //   console.log("Issue " + query);
        //   console.log(idx);
        return;
      } else {
        cc.count += 1;
      }
    });

    type ClusterReport = {
      count: number;
      field: string;
      normalized_count: number;
    };

    const reportedClusters: ClusterReport[] = Array.from(clusterCounts.values())
      .map((cc: ClusterCount) => {
        const count = cc.reduce((acc, next) => acc + next.count, 0);
        const normalized_count = count / queries.length;
        // most frequent query
        const field = cc.sort((a, b) => b.count - a.count)[0].query;

        return { count, field, normalized_count };
      })
      .sort((a, b) => b.count - a.count);

    return new DataFrame(reportedClusters);
  } else {
    // TODO this is ugly
    throw new Error("Cache not set");
  }
};

/**
 *
 * @param dataset
 * @param m0 focus period : m0
 * @param m1 reference period (m0 - 1)
 * @param m2 two months ago (m0 - 2)
 * @param reportId
 * @param popularityType
 * @param cache
 */
const analyse = (
  dataset: IDataFrame,
  m0: string[],
  m1: string[],
  m2: string[],
  reportId: string,
  popularityType: PopularityTypeString,
  cache: Option<Cache> = none
): PopularityReport[] => {
  // if query popoularity, cache is required
  if (popularityType == PopularityType[PopularityType.QUERY] && cache == none) {
    throw new Error("Cache required for query popularity report");
  }

  const visits = getVisits(dataset);

  // FIXME for unkwown reason this doesn't work
  /*
  console.log(
    visits
      .select((visit) => toUniqueViews(visit))
      .inflate()
      .take(2)
      .toString()
  );
  */

  // so we use toArray for now
  const grouping =
    popularityType == PopularityType[PopularityType.QUERY]
      ? toUniqueSearches
      : toUniqueViews;
  const uniqueViews = DataFrame.concat(
    visits.select((visit) => grouping(visit)).toArray()
  );

  const idxUniqueViews = uniqueViews.withIndex(
    Array.from(Array(uniqueViews.count()).keys())
  );

  // clean views
  const noError = (action: any) =>
    ![
      "https://code.travail.gouv.fr/",
      "https://code.travail.gouv.fr/?xtor=ES-29-[BIE_202_20200130]-20200130-[https://code.travail.gouv.fr/]",
      "https://code.travail.gouv.fr/droit-du-travail",
    ].includes(action.url);

  // TODO change this
  let filter = "";
  if (popularityType == PopularityType[PopularityType.CONVENTION]) {
    filter = "convention-collective/";
  }

  const filterUrl = (action: any) =>
    filter ? action.url.includes(filter) : true;

  const filteredVisitViews = idxUniqueViews.where(noError).where(filterUrl);

  const removeAnchor = (url: string) => {
    return url.split("#")[0];
  };

  const cleanedViews = filteredVisitViews.transformSeries({
    url: (u) => urlToPath(removeAnchor(u)),
  });

  const focus = cleanedViews.where((a) => m0.includes(a.logfile));
  const reference = cleanedViews.where((a) => m1.includes(a.logfile));
  const m2Data = cleanedViews.where((a) => m2.includes(a.logfile));

  const m0Start = focus.getSeries("timestamp").min();
  const m1Start = reference.getSeries("timestamp").min();
  const m2Start = m2Data.getSeries("timestamp").min();

  let m0Counts, m1Counts, m2Counts;
  if (popularityType != PopularityType[PopularityType.QUERY]) {
    m0Counts = countURLs(focus);
    m1Counts = countURLs(reference);
    m2Counts = countURLs(m2Data);
  } else {
    m0Counts = countQueries(focus, cache);
    m1Counts = countQueries(reference, cache);
    m2Counts = countQueries(m2Data, cache);
  }

  const reports = computeReports(
    m0Counts,
    m1Counts,
    m2Counts,

    m0Start,
    m1Start,
    m2Start,

    reportId,
    reportType(popularityType)
  );

  return reports;
};

export { analyse, reportType };
