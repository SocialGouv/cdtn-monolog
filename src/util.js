export const SERVICE_URL = "https://code.travail.gouv.fr/";

export const urlToPath = (url) => {
  if (url.startsWith(SERVICE_URL)) {
    return url.slice(SERVICE_URL.length);
  } else {
    return url;
  }
};

export const actionTypes = {
  visitType: "visit_content",
  selectSuggestionType: "select_suggestion",
};
