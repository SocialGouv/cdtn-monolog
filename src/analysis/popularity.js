import * as dataForge from "data-forge";
import * as util from "../util";
import * as datasetUtil from "../dataset";

const reportType = "popularity";

const analyse = (dataset, proportion, reportId) => {
  const visits = datasetUtil.getVisits(dataset);

  // FIXME for unkwown reason this doesn't work
  /*
  console.log(
    visits
      .select((visit) => toUniqueViews(visit))
      .inflate()
      .take(2)
      .toString()
  );
  */

  // so we use toArray for now
  const uniqueViews = dataForge.DataFrame.concat(
    visits.select((visit) => datasetUtil.toUniqueViews(visit)).toArray()
  );

  // clean views
  const noError = (action) =>
    ![
      "https://code.travail.gouv.fr/",
      "https://code.travail.gouv.fr/?xtor=ES-29-[BIE_202_20200130]-20200130-[https://code.travail.gouv.fr/]",
      "https://code.travail.gouv.fr/droit-du-travail",
    ].includes(action.url);

  const filteredVisitViews = uniqueViews.where(noError);

  const removeAnchor = (url) => {
    return url.split("#")[0];
  };

  const cleanedViews = filteredVisitViews.transformSeries({
    url: (u) => util.urlToPath(removeAnchor(u)),
  });

  const dates = uniqueViews.deflate((r) => r.timestamp);
  const start = dates.min();
  const end = dates.max();
  const refDate = Math.floor(start + (1 - proportion) * (end - start));

  const afterRef = (a) => a.timestamp > refDate;

  const focus = cleanedViews.where(afterRef);
  const reference = cleanedViews.where((a) => !afterRef(a));

  const countURLs = (dataframe) => {
    const counts = dataframe
      .deflate((a) => a.url)
      .groupBy((value) => value)
      .select((group) => {
        return {
          url: group.first(),
          count: group.count(),
        };
      })
      .inflate()
      .orderByDescending((r) => r.count);

    const sum = counts.deflate((r) => r.count).sum();
    const normalizedCounts = counts.withSeries({
      normalized_count: (df) =>
        df.deflate((row) => row.count).select((count) => count / sum),
    });
    return normalizedCounts.setIndex("url");
  };

  const refCounts = countURLs(reference);
  const focusCounts = countURLs(focus);

  // FIXME use outer join to handle missing values (e.g. additions)
  const joined = refCounts.join(
    focusCounts,
    (left) => left.url,
    (right) => right.url,
    (left, right) => {
      return {
        url: left.url,
        ref_norm_count: left.normalized_count,
        focus_norm_count: right.normalized_count,
        ref_count: left.count,
        focus_count: right.count,
      };
    }
  );

  const nContent = 40;

  const diff = joined
    .generateSeries({
      diff: (row) => row.focus_norm_count - row.ref_norm_count,
    })
    .generateSeries({
      abs_diff: (row) => Math.abs(row.diff),
    })
    .orderByDescending((r) => r.abs_diff)
    .take(nContent);

  return diff.toArray().map((doc) => ({
    doc,
    start: start,
    end: end,
    pivot: refDate,
    reportType,
    reportId,
  }));
};

export { analyse, reportType };
