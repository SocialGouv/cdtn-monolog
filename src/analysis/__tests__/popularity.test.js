import * as Popularity from "../popularity";
import * as util from "../../util";
import { DataFrame } from "data-forge";

describe("Popularity", () => {
  // FIXME : move as duplicated with covisit
  const toVisitAction = (uvi, url, timestamp) => ({
    uvi,
    url,
    timestamp,
    type: util.actionTypes.visitType,
  });
  const url = (i) => `url${i}`;

  // one day in seconds
  const sDay = 60 * 60 * 24;

  // get date for today - n days
  const getTimestamp = (n) =>
    Math.floor(new Date(new Date().getTime() - n * sDay).getTime());

  const generate = (a) =>
    new DataFrame(a.map(([uvi, u, ts]) => toVisitAction(uvi, url(u), ts)));

  // takes triples and generate data : [1,3,2] =>  visit id 1 at date 2 days ago / content viewed 3
  const generateWithDays = (a) =>
    generate(a.map(([uvi, u, n]) => [uvi, u, getTimestamp(n)]));

  it("should use unique views per visit", () => {
    const actions = generateWithDays([
      [1, 1, 1],
      [1, 1, 1],
      [1, 2, 1],

      [2, 1, 1],
      [2, 2, 1],

      [3, 1, 10],
      [3, 1, 10],
      [3, 2, 10],
    ]);
    // console.log(actions.toString());

    // we ensure that count are ok for focus and reference periods for url1
    const res = Popularity.analyse(actions, 0.2, 42).filter(
      (r) => r.doc.url == url(1)
    )[0].doc;
    expect(res.ref_count).toBe(1);
    expect(res.focus_count).toBe(2);
  });

  it("should respect proportion between reference and focus period", () => {
    const actions = generateWithDays([
      // visit 11 = visit 1 for days -1
      // visit 21 = visit 2 for days -1
      [11, 1, 1],
      [11, 2, 1],
      [21, 1, 1],
      [21, 2, 1],

      // visit 12 = visit 1 for days -2
      [12, 1, 2],
      [12, 2, 2],
      [22, 1, 2],
      [22, 2, 2],

      [13, 1, 3],
      [13, 2, 3],
      [23, 1, 3],
      [23, 2, 3],
    ]);

    // we ensure that count are ok for focus and reference periods for url1
    const reports1 = Popularity.analyse(actions, 0.3, 43);
    const res1 = reports1.filter((r) => r.doc.url == url(1))[0].doc;
    expect(res1.ref_count).toBe(4);
    expect(res1.focus_count).toBe(2);
    expect(res1).toMatchSnapshot();

    // and using a different split
    const reports2 = Popularity.analyse(actions, 0.7, 44);
    const res2 = reports2.filter((r) => r.doc.url == url(1))[0].doc;
    expect(res2.ref_count).toBe(2);
    expect(res2.focus_count).toBe(4);
    expect(res2).toMatchSnapshot();
  });

  it("should generate report as expected, including dates", () => {
    const end = Math.floor(1588608172467 / 1000);
    const start = end - 10 * sDay;
    const actions = generate([
      [21, 1, end],
      [12, 1, end - sDay],
      [22, 1, end - sDay],
      [13, 1, start],
      [23, 1, start],
    ]);

    // we ensure that count are ok for focus and reference periods for url1
    const report1 = Popularity.analyse(actions, 0.3, 45)[0];
    expect(report1).toMatchSnapshot();
    expect(report1.end).toBe(end);
    expect(report1.start).toBe(start);
    expect(report1.pivot).toBe(start + 0.7 * (end - start));
  });
});
