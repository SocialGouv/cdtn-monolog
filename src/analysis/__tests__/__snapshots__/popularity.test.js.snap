// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popularity should generate report as expected, including dates 1`] = `
Object {
  "doc": Object {
    "abs_diff": 0,
    "diff": 0,
    "focus_count": 3,
    "focus_norm_count": 1,
    "ref_count": 2,
    "ref_norm_count": 1,
    "url": "url1",
  },
  "end": 1588608172,
  "pivot": 1588348972,
  "reportId": 45,
  "reportType": "popularity",
  "start": 1587744172,
}
`;

exports[`Popularity should respect proportion between reference and focus period 1`] = `
Object {
  "abs_diff": 0,
  "diff": 0,
  "focus_count": 2,
  "focus_norm_count": 0.5,
  "ref_count": 4,
  "ref_norm_count": 0.5,
  "url": "url1",
}
`;

exports[`Popularity should respect proportion between reference and focus period 2`] = `
Object {
  "abs_diff": 0,
  "diff": 0,
  "focus_count": 4,
  "focus_norm_count": 0.5,
  "ref_count": 2,
  "ref_norm_count": 0.5,
  "url": "url1",
}
`;
